pandas/tests/io/json/test_pandas.py

```diff
@@ -470,101 +471,99 @@ def _check_all_orients(
 
-    def test_frame_from_json_bad_data(self):
-        with pytest.raises(ValueError, match="Expected object or value"):
-            read_json(StringIO('{"key":b:a:d}'))
-
-        # too few indices
-        json = StringIO(
-            '{"columns":["A","B"],'
-            '"index":["2","3"],'
-            '"data":[[1.0,"1"],[2.0,"2"],[null,"3"]]}'
-        )
-        msg = r"Shape of passed values is \(3, 2\), indices imply \(2, 2\)"
+    @pytest.mark.parametrize(
+        "data,msg,orient",
+        [
+            (StringIO('{"key":b:a:d}'), "Expected object or value", "columns"),
+            # too few indices
+            (
+                StringIO(
+                    '{"columns":["A","B"],'
+                    '"index":["2","3"],'
+                    '"data":[[1.0,"1"],[2.0,"2"],[null,"3"]]}'
+                ),
+                r"Shape of passed values is \(3, 2\), indices imply \(2, 2\)",
+                "split",
+            ),
+            # too many columns
+            (
+                StringIO(
+                    '{"columns":["A","B","C"],'
+                    '"index":["1","2","3"],'
+                    '"data":[[1.0,"1"],[2.0,"2"],[null,"3"]]}'
+                ),
+                "3 columns passed, passed data had 2 columns",
+                "split",
+            ),
+            # bad key
+            (
+                StringIO(
+                    '{"badkey":["A","B"],'
+                    '"index":["2","3"],'
+                    '"data":[[1.0,"1"],[2.0,"2"],[null,"3"]]}'
+                ),
+                r"unexpected key\(s\): badkey",
+                "split",
+            ),
+        ],
+    )
+    def test_frame_from_json_bad_data_raises(self, data, msg, orient):
         with pytest.raises(ValueError, match=msg):
-            read_json(json, orient="split")
+            read_json(data, orient=orient)
 
-        # too many columns
-        json = StringIO(
-            '{"columns":["A","B","C"],'
-            '"index":["1","2","3"],'
-            '"data":[[1.0,"1"],[2.0,"2"],[null,"3"]]}'
+    @pytest.mark.parametrize("dtype", [True, False])
+    @pytest.mark.parametrize("convert_axes", [True, False])
+    @pytest.mark.parametrize("numpy", [True, False])
+    def test_frame_from_json_missing_data(self, orient, convert_axes, numpy, dtype):
+        num_df = DataFrame([[1, 2], [4, 5, 6]])
+        result = read_json(
+            num_df.to_json(orient=orient),
+            orient=orient,
+            convert_axes=convert_axes,
+            dtype=dtype,
         )
-        msg = "3 columns passed, passed data had 2 columns"
-        with pytest.raises(ValueError, match=msg):
-            read_json(json, orient="split")
+        assert np.isnan(result.iloc[0, 2])
 
-        # bad key
-        json = StringIO(
-            '{"badkey":["A","B"],'
-            '"index":["2","3"],'
-            '"data":[[1.0,"1"],[2.0,"2"],[null,"3"]]}'
+        obj_df = DataFrame([["1", "2"], ["4", "5", "6"]])
```

Comment:
it looks like this makes two changes vs test_frame_from_json_nones on 506:

- using iloc[0, 2] instead of case-specific indexing (seems fine unless testing the index/column types was important)
- adds two cases with convert_axes=dtype=True.  Seems like a strict improvement, unless there is some non-obvious reason those two were excluded.

Conditional on these being intentional, this func LGTM

Reply:
Yep sorry should have commented. The `.iloc` is certainly a change to make it more idiomatic and the `convert_axes` parametrization is to improve test coverage (not sure why previous implementation only tested for False)